### PR TITLE
Migrate authorization filters to authorization policies

### PR DESCRIPTION
**Pull request by Shazwazza**
Migrate authorization filters to authorization policies. Not all implementations of IAuthorizationFilter have been replaced since not all are necessary to be migrated such as InstallAuthorizeAttribute. But all filters that deal with back office authorization have been migrated.

This creates a different policy (name) per access rule. This does mean there are a lot of rules listed in the AuthorizationPolicies but it does mean we also create all of the rules up front. There is an alternative to this which would be to revert back to custom IActionFilter attributes that uses the IAuthorizationService directly and invoke custom authorization policies via [resource authorization](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/resourcebased?view=aspnetcore-5.0) and dynamically pass in the requirements/resources based on the parameters specified in those attributes. Basically it's either we define the rules/policies in one place, or they are dispersed within attributes in controllers. My vote is to leave them all declared as policies like in this PR since we can then target them just by name and they are managed globally.

## Testing

* Ensure the back office works as expected
* Create user(s) that have different access levels and ensure authorization works as expected
* There's a lot of scenarios, not expecting that all are tested - this is mostly code review since the code for the authorization itself hasn't changed

### CODE REVIEW

File: src/Umbraco.Web.BackOffice/Controllers/RelationController.cs [hunk 1]
```diff
@@ -56,20 +58,5 @@ public IEnumerable<RelationDisplay> GetByChildId(int childId, string relationTyp
             return _umbracoMapper.MapEnumerable<IRelation, RelationDisplay>(relations);
         }
 
-        [HttpDelete]
```
Comment (Shazwazza) [inline src/Umbraco.Web.BackOffice/Controllers/RelationController.cs:59]: This method is not used and shouldn't exist so that's why it's deleted

**Comment by bergmania** [comment]
@Shazwazza are you aware of the failing tests?

File: src/Umbraco.Web.BackOffice/Authorization/SectionHandler.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                context.Succeed(requirement);
+            }
+            else
+            {
+                context.Fail();
+            }
+
+            return Task.CompletedTask;
+        }
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Authorization/SectionHandler.cs:36]: I will look into it.

File: src/Umbraco.Web.BackOffice/Controllers/PreviewController.cs [hunk 3]
```diff
@@ -64,7 +66,7 @@ public class PreviewController : Controller
             _viewEngines = viewEngines;
         }
 
-        [UmbracoBackOfficeAuthorize(redirectToUmbracoLogin: true, requireApproval : false)]
+        [Authorize(Policy = AuthorizationPolicies.BackOfficeAccessWithoutApproval)]
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Controllers/PreviewController.cs:69]: It's not needed, redirection happens as part of the authentication, this was something based on very old things back when there was webforms forms authentication.

File: src/Umbraco.Web.Website/Controllers/UmbracoAuthorizedController.cs [hunk 4]
```diff
@@ -9,10 +11,11 @@ namespace Umbraco.Web.Mvc
     /// This controller essentially just uses a global UmbracoAuthorizeAttribute, inheritors that require more granular control over the
     /// authorization of each method can use this attribute instead of inheriting from this controller.
     /// </remarks>
-    [UmbracoBackOfficeAuthorize]
+    [Authorize(Policy = AuthorizationPolicies.BackOfficeAccess)]
     [DisableBrowserCache]
     public abstract class UmbracoAuthorizedController : UmbracoController
     {
+        // TODO: This controller is not used at all, is there a need for this controller?
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.Website/Controllers/UmbracoAuthorizedController.cs:18]: I will remove

**Comment by Shazwazza** [comment]
@bergmania no i didn't know there was failing tests, i'll look

File: src/Umbraco.Web.BackOffice/Authorization/SectionHandler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Authorization/SectionHandler.cs]: have updated

**Comment by Shazwazza** [comment]
@bergmania should all be good, integration tests are passing on my machine now (just waiting to see if they pass on the server)

**Comment by bergmania** [comment]
💪 Code changes looks fine. If build server is happy, I will do a short manual test, and merge if nothing comes up

**Comment by bergmania** [comment]
I got this error 
```
[14:24:16 ERR] An unhandled exception has occurred while executing the request.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Umbraco.Web.BackOffice.Authorization.MustSatisfyRequirementAuthorizationHandler`1.HandleRequirementAsync(AuthorizationHandlerContext context, T requirement) in D:\Projects\Umbraco-CMS-netcore\src\Umbraco.Web.BackOffice\Authorization\MustSatisfyRequirementAuthorizationHandler.cs:line 18
   at Microsoft.AspNetCore.Authorization.AuthorizationHandler`1.HandleAsync(AuthorizationHandlerContext context)
   at Microsoft.AspNetCore.Authorization.DefaultAuthorizationService.AuthorizeAsync(ClaimsPrincipal user, Object resource, IEnumerable`1 requirements)
   at Microsoft.AspNetCore.Authorization.Policy.PolicyEvaluator.AuthorizeAsync(AuthorizationPolicy policy, AuthenticateResult authenticationResult, HttpContext context, Object resource)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at StackExchange.Profiling.MiniProfilerMiddleware.Invoke(HttpContext context) in C:\projects\dotnet\src\MiniProfiler.AspNetCore\MiniProfilerMiddleware.cs:line 121
   at Umbraco.Web.Common.Middleware.UmbracoRequestMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in D:\Projects\Umbraco-CMS-netcore\src\Umbraco.Web.Common\Middleware\UmbracoRequestMiddleware.cs:line 80
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

```

The error comes from UserGroupHandler returning null .. I will change to false..

File: src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            _backofficeSecurityAccessor = backofficeSecurityAccessor;
+        }
+
+        protected override Task<bool> IsAuthorized(AuthorizationHandlerContext context, UserGroupRequirement requirement)
+        {
+            var currentUser = _backofficeSecurityAccessor.BackOfficeSecurity.CurrentUser;
+
+            var queryString = _httpContextAcessor.HttpContext?.Request.Query;
+            if (queryString == null)
+                return null;
```
Comment (bergmania) [inline src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs:45]: Returning null leads to null ref issues..

File: src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+            var ids = queryString.Where(x => x.Key == requirement.QueryStringName).ToArray();
+            if (ids.Length == 0)
+                return null;
+
+            var intIds = ids.Select(x => x.Value.TryConvertTo<int>()).Where(x => x.Success).Select(x => x.Result).ToArray();
```
Comment (bergmania) [inline src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs:51]: `TryConvertTo` don't support `StringValues`. We need to do `ToString()` or even better, let `TryConvertTo` support `StringValues`

File: src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs [hunk 7]
```diff
@@ -164,7 +166,7 @@ public IEnumerable<UserGroupBasic> GetUserGroups(bool onlyCurrentUserGroups = tr
         /// Return a user group
         /// </summary>
         /// <returns></returns>
-        [UserGroupAuthorization("id")]
+        [Authorize(Policy = AuthorizationPolicies.UserBelongsToUserGroupInRequest)]
```
Comment (bergmania) [inline src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs:169]: The querystring name is `id` and not `userGroupIds` like below, so out auth is not hit correct

**Review by bergmania — CHANGES_REQUESTED** [review]
I found some other issues when creating a User Group.

File: src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs (hunk 5 above)
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs:45]: Doh, that's a refactor mistake

File: src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs (hunk 6 above)
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Authorization/UserGroupHandler.cs]: Fixed here and in AdminUsersHandler

File: src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs (hunk 7 above)
Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs:169]: Yes, the policy looks for both query strings

```cs
            options.AddPolicy(AuthorizationPolicies.UserBelongsToUserGroupInRequest, policy =>
            {
                policy.AuthenticationSchemes.Add(backOfficeAuthenticationScheme);
                policy.Requirements.Add(new UserGroupRequirement());
                policy.Requirements.Add(new UserGroupRequirement("userGroupIds"));
            });
```

Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs:169]: Both requirements must be met for it to succeed. When the query string doesn't match a value it returns true to satisfy the requirement but one of the requirements will match the query string.

Comment (Shazwazza) [reply src/Umbraco.Web.BackOffice/Controllers/UserGroupsController.cs:169]: Have tested and this is working

**Pull request merged**